### ISSUE TITLE
💅 noUnnecessaryConditions false positives

### ISSUE THREAD

**Issue by CMTegner** [biomejs/biome#10704]
### Environment information

<details>

```bash

CLI:
  Version:                      2.5.0
  Color support:                true

Platform:
  CPU Architecture:             aarch64
  OS:                           macos

Environment:
  BIOME_DISTRIBUTION:                npm
  BIOME_LOG_PATH:                    unset
  BIOME_LOG_PREFIX_NAME:             unset
  BIOME_LOG_LEVEL:                   unset
  BIOME_LOG_KIND:                    unset
  BIOME_CONFIG_PATH:                 unset
  BIOME_THREADS:                     unset
  BIOME_WATCHER_KIND:                unset
  BIOME_WATCHER_POLLING_INTERVAL:    unset
  NO_COLOR:                     unset
  TERM:                         tmux-256color
  JS_RUNTIME_VERSION:           v24.15.0
  JS_RUNTIME_NAME:              node
  NODE_PACKAGE_MANAGER:         unset

Biome Configuration:
  Status:                       Loaded successfully.
  Path:                         biome.json
  Formatter enabled:            true
  Linter enabled:               true
  Assist enabled:               true
  VCS enabled:                  true
  HTML full support enabled:    unset

Linter:
  JavaScript enabled:           unset
  JSON enabled:                 unset
  CSS enabled:                  unset
  GraphQL enabled:              unset
  Recommended:                  unset
  Enabled rules:
    a11y/noAccessKey
    a11y/noAriaHiddenOnFocusable
    a11y/noAriaUnsupportedElements
    a11y/noAutofocus
    a11y/noDistractingElements
    a11y/noHeaderScope
    a11y/noInteractiveElementToNoninteractiveRole
    a11y/noLabelWithoutControl
    a11y/noNoninteractiveElementInteractions
    a11y/noNoninteractiveElementToInteractiveRole
    a11y/noNoninteractiveTabindex
    a11y/noPositiveTabindex
    a11y/noRedundantAlt
    a11y/noRedundantRoles
    a11y/noStaticElementInteractions
    a11y/noSvgWithoutTitle
    a11y/useAltText
    a11y/useAnchorContent
    a11y/useAriaActivedescendantWithTabindex
    a11y/useAriaPropsForRole
    a11y/useAriaPropsSupportedByRole
    a11y/useButtonType
    a11y/useFocusableInteractive
    a11y/useGenericFontNames
    a11y/useHeadingContent
    a11y/useHtmlLang
    a11y/useIframeTitle
    a11y/useKeyWithClickEvents
    a11y/useKeyWithMouseEvents
    a11y/useMediaCaption
    a11y/useSemanticElements
    a11y/useValidAnchor
    a11y/useValidAriaProps
    a11y/useValidAriaRole
    a11y/useValidAriaValues
    a11y/useValidAutocomplete
    a11y/useValidLang
    complexity/noAdjacentSpacesInRegex
    complexity/noArguments
    complexity/noBannedTypes
    complexity/noCommaOperator
    complexity/noDivRegex
    complexity/noEmptyTypeParameters
    complexity/noExcessiveNestedTestSuites
    complexity/noExtraBooleanCast
    complexity/noFlatMapIdentity
    complexity/noForEach
    complexity/noImplicitCoercions
    complexity/noImportantStyles
    complexity/noRedundantDefaultExport
    complexity/noStaticOnlyClass
    complexity/noThisInStatic
    complexity/noUselessCatch
    complexity/noUselessCatchBinding
    complexity/noUselessConstructor
    complexity/noUselessContinue
    complexity/noUselessEmptyExport
    complexity/noUselessEscapeInRegex
    complexity/noUselessFragments
    complexity/noUselessLabel
    complexity/noUselessLoneBlockStatements
    complexity/noUselessRename
    complexity/noUselessReturn
    complexity/noUselessStringConcat
    complexity/noUselessStringRaw
    complexity/noUselessSwitchCase
    complexity/noUselessTernary
    complexity/noUselessThisAlias
    complexity/noUselessTypeConstraint
    complexity/noUselessUndefined
    complexity/noUselessUndefinedInitialization
    complexity/noVoid
    complexity/useArrayFind
    complexity/useArrowFunction
    complexity/useDateNow
    complexity/useFlatMap
    complexity/useIndexOf
    complexity/useLiteralKeys
    complexity/useMaxParams
    complexity/useNumericLiterals
    complexity/useOptionalChain
    complexity/useRegexLiterals
    complexity/useSimpleNumberKeys
    complexity/useSimplifiedLogicExpression
    complexity/useWhile
    correctness/noChildrenProp
    correctness/noConstAssign
    correctness/noConstantCondition
    correctness/noConstantMathMinMaxClamp
    correctness/noConstructorReturn
    correctness/noDuplicateArgumentNames
    correctness/noDuplicateEnumValueNames
    correctness/noDuplicateInputFieldNames
    correctness/noEmptyCharacterClassInRegex
    correctness/noEmptyPattern
    correctness/noGlobalDirnameFilename
    correctness/noGlobalObjectCalls
    correctness/noInnerDeclarations
    correctness/noInvalidBuiltinInstantiation
    correctness/noInvalidConstructorSuper
    correctness/noInvalidDirectionInLinearGradient
    correctness/noInvalidGridAreas
    correctness/noInvalidPositionAtImportRule
    correctness/noInvalidUseBeforeDeclaration
    correctness/noMissingVarFunction
    correctness/noNestedComponentDefinitions
    correctness/noNonoctalDecimalEscape
    correctness/noPrecisionLoss
    correctness/noProcessGlobal
    correctness/noReactPropAssignments
    correctness/noRenderReturnValue
    correctness/noSelfAssign
    correctness/noSetterReturn
    correctness/noStringCaseMismatch
    correctness/noSwitchDeclarations
    correctness/noUndeclaredDependencies
    correctness/noUndeclaredVariables
    correctness/noUnknownFunction
    correctness/noUnknownMediaFeatureName
    correctness/noUnknownProperty
    correctness/noUnknownPseudoClass
    correctness/noUnknownPseudoElement
    correctness/noUnknownTypeSelector
    correctness/noUnknownUnit
    correctness/noUnmatchableAnbSelector
    correctness/noUnreachable
    correctness/noUnreachableSuper
    correctness/noUnsafeFinally
    correctness/noUnsafeOptionalChaining
    correctness/noUnusedFunctionParameters
    correctness/noUnusedImports
    correctness/noUnusedInstantiation
    correctness/noUnusedLabels
    correctness/noUnusedPrivateClassMembers
    correctness/noUnusedVariables
    correctness/noVoidElementsWithChildren
    correctness/noVoidTypeReturn
    correctness/useExhaustiveDependencies
    correctness/useGraphqlNamedOperations
    correctness/useHookAtTopLevel
    correctness/useImageSize
    correctness/useImportExtensions
    correctness/useIsNan
    correctness/useJsonImportAttributes
    correctness/useJsxKeyInIterable
    correctness/useLoneAnonymousOperation
    correctness/useParseIntRadix
    correctness/useSingleJsDocAsterisk
    correctness/useValidForDirection
    correctness/useValidTypeof
    correctness/useYield
    performance/noAccumulatingSpread
    performance/noBarrelFile
    performance/noDelete
    performance/noDynamicNamespaceImportAccess
    performance/noJsxPropsBind
    performance/noNamespaceImport
    performance/noReExportAll
    performance/useGoogleFontPreconnect
    performance/useTopLevelRegex
    security/noBlankTarget
    security/noDangerouslySetInnerHtmlWithChildren
    security/noGlobalEval
    security/noScriptUrl
    style/noCommonJs
    style/noDescendingSpecificity
    style/noEnum
    style/noExcessiveClassesPerFile
    style/noExportedImports
    style/noHexColors
    style/noInferrableTypes
    style/noMultiAssign
    style/noMultilineString
    style/noNamespace
    style/noNegationElse
    style/noNonNullAssertion
    style/noParameterAssign
    style/noParameterProperties
    style/noProcessEnv
    style/noRestrictedGlobals
    style/noShoutyConstants
    style/noSubstr
    style/noUnusedTemplateLiteral
    style/noUselessElse
    style/noValueAtRule
    style/noYodaExpression
    style/useArrayLiterals
    style/useAsConstAssertion
    style/useAtIndex
    style/useBlockStatements
    style/useCollapsedElseIf
    style/useCollapsedIf
    style/useComponentExportOnlyModules
    style/useConsistentArrayType
    style/useConsistentArrowReturn
    style/useConsistentBuiltinInstantiation
    style/useConsistentCurlyBraces
    style/useConsistentGraphqlDescriptions
    style/useConsistentMemberAccessibility
    style/useConsistentMethodSignatures
    style/useConsistentObjectDefinitions
    style/useConsistentTypeDefinitions
    style/useConst
    style/useDefaultParameterLast
    style/useDefaultSwitchClause
    style/useDeprecatedReason
    style/useDestructuring
    style/useEnumInitializers
    style/useExplicitLengthCheck
    style/useExponentiationOperator
    style/useExportType
    style/useExportsLast
    style/useFilenamingConvention
    style/useForOf
    style/useFragmentSyntax
    style/useGlobalThis
    style/useGraphqlNamingConvention
    style/useGroupedAccessorPairs
    style/useImportType
    style/useInputName
    style/useLiteralEnumMembers
    style/useNodeAssertStrict
    style/useNodejsImportProtocol
    style/useNumberNamespace
    style/useNumericSeparators
    style/useObjectSpread
    style/useReactFunctionComponents
    style/useReadonlyClassProperties
    style/useSelfClosingElements
    style/useShorthandAssign
    style/useShorthandFunctionType
    style/useSingleVarDeclarator
    style/useSpreadOverApply
    style/useSymbolDescription
    style/useTemplate
    style/useThrowNewError
    style/useThrowOnlyError
    style/useTrimStartEnd
    style/useUnifiedTypeSignatures
    suspicious/noAlert
    suspicious/noApproximativeNumericConstant
    suspicious/noArrayIndexKey
    suspicious/noAssignInExpressions
    suspicious/noAsyncPromiseExecutor
    suspicious/noBiomeFirstException
    suspicious/noBitwiseOperators
    suspicious/noCatchAssign
    suspicious/noClassAssign
    suspicious/noCommentText
    suspicious/noCompareNegZero
    suspicious/noConfusingLabels
    suspicious/noConfusingVoidType
    suspicious/noConsole
    suspicious/noConstEnum
    suspicious/noConstantBinaryExpressions
    suspicious/noControlCharactersInRegex
    suspicious/noDebugger
    suspicious/noDeprecatedImports
    suspicious/noDocumentCookie
    suspicious/noDoubleEquals
    suspicious/noDuplicateAtImportRules
    suspicious/noDuplicateCase
    suspicious/noDuplicateClassMembers
    suspicious/noDuplicateCustomProperties
    suspicious/noDuplicateDependencies
    suspicious/noDuplicateElseIf
    suspicious/noDuplicateEnumValues
    suspicious/noDuplicateFields
    suspicious/noDuplicateFontNames
    suspicious/noDuplicateGraphqlOperationName
    suspicious/noDuplicateJsxProps
    suspicious/noDuplicateObjectKeys
    suspicious/noDuplicateParameters
    suspicious/noDuplicateProperties
    suspicious/noDuplicateSelectorsKeyframeBlock
    suspicious/noDuplicateTestHooks
    suspicious/noDuplicatedSpreadProps
    suspicious/noEmptyBlock
    suspicious/noEmptyBlockStatements
    suspicious/noEmptyInterface
    suspicious/noEmptySource
    suspicious/noEqualsToNull
    suspicious/noEvolvingTypes
    suspicious/noExplicitAny
    suspicious/noExportsInTest
    suspicious/noExtraNonNullAssertion
    suspicious/noFallthroughSwitchClause
    suspicious/noFocusedTests
    suspicious/noForIn
    suspicious/noFunctionAssign
    suspicious/noGlobalAssign
    suspicious/noGlobalIsFinite
    suspicious/noGlobalIsNan
    suspicious/noImplicitAnyLet
    suspicious/noImportAssign
    suspicious/noImportCycles
    suspicious/noImportantInKeyframe
    suspicious/noIrregularWhitespace
    suspicious/noLabelVar
    suspicious/noMisleadingCharacterClass
    suspicious/noMisleadingInstantiator
    suspicious/noMisplacedAssertion
    suspicious/noMisrefactoredShorthandAssign
    suspicious/noNestedPromises
    suspicious/noNonNullAssertedOptionalChain
    suspicious/noOctalEscape
    suspicious/noParametersOnlyUsedInRecursion
    suspicious/noProto
    suspicious/noPrototypeBuiltins
    suspicious/noQuickfixBiome
    suspicious/noReactForwardRef
    suspicious/noRedeclare
    suspicious/noRedundantUseStrict
    suspicious/noReturnAssign
    suspicious/noSelfCompare
    suspicious/noShadow
    suspicious/noShadowRestrictedNames
    suspicious/noShorthandPropertyOverrides
    suspicious/noSkippedTests
    suspicious/noSparseArray
    suspicious/noSuspiciousSemicolonInJsx
    suspicious/noTemplateCurlyInString
    suspicious/noThenProperty
    suspicious/noTsIgnore
    suspicious/noUnassignedVariables
    suspicious/noUnknownAtRules
    suspicious/noUnsafeDeclarationMerging
    suspicious/noUnsafeNegation
    suspicious/noUnusedExpressions
    suspicious/noUselessEscapeInString
    suspicious/noUselessRegexBackrefs
    suspicious/noVar
    suspicious/noWith
    suspicious/useAdjacentOverloadSignatures
    suspicious/useArraySortCompare
    suspicious/useAwait
    suspicious/useBiomeIgnoreFolder
    suspicious/useDefaultSwitchClauseLast
    suspicious/useDeprecatedDate
    suspicious/useErrorMessage
    suspicious/useGetterReturn
    suspicious/useGoogleFontDisplay
    suspicious/useIsArray
    suspicious/useIterableCallbackReturn
    suspicious/useNamespaceKeyword
    suspicious/useNumberToFixedDigitsArgument
    suspicious/useRequiredScripts
    suspicious/useStaticResponseMethods

Workspace:
  Open Documents:               0

```

</details>

### Rule name

noUnnecessaryConditions

### Playground link

https://biomejs.dev/playground/?lintRules=noUnnecessaryConditions&tab=formatter&pane=Diagnostics&code=dAB5AHAAZQAgAE0AbwB2AGkAZQAgAD0AIAB7AAoACQBjAGgAYQByAGEAYwB0AGUAcgBzAD8AOgAgAEEAcgByAGEAeQA8AHsACgAJAAkAaQBzAEoAZQBkAGkAOgAgAGIAbwBvAGwAZQBhAG4AOwAKAAkACQBuAGEAbQBlADoAIABzAHQAcgBpAG4AZwA7AAoACQB9AD4AOwAKAH0AOwAKAAoAZQB4AHAAbwByAHQAIABmAHUAbgBjAHQAaQBvAG4AIABnAGUAdABKAGUAZABpAE4AYQBtAGUAcwAoAG0AbwB2AGkAZQA6ACAATQBvAHYAaQBlACkAOgAgAEEAcgByAGEAeQA8AHMAdAByAGkAbgBnAD4AIAB7AAoACQBjAG8AbgBzAHQAIABqAGUAZABpAHMAIAA9ACAAbQBvAHYAaQBlAC4AYwBoAGEAcgBhAGMAdABlAHIAcwA%2FAC4AZgBpAGwAdABlAHIAKAAoAHUAKQAgAD0APgAgAHUALgBpAHMASgBlAGQAaQApACAAPwA%2FACAAWwBdADsACgAJAHIAZQB0AHUAcgBuACAAagBlAGQAaQBzAC4AbQBhAHAAKAAoAHUAKQAgAD0APgAgAHUALgBuAGEAbQBlACkAOwAKAH0ACgAKAGMAbwBuAHMAdAAgAEYAQQBDAFQASQBPAE4AUwAgAD0AIABbAAoACQB7ACAAbgBhAG0AZQA6ACAAIgBSAEUAUABVAEIATABJAEMAIgAsACAAaQBkADoAIAAxACAAfQAsAAoACQB7ACAAbgBhAG0AZQA6ACAAIgBFAE0AUABJAFIARQAiACwAIABpAGQAOgAgADIAIAB9ACwACgAJAHsAIABuAGEAbQBlADoAIAAiAEEATABMAEkAQQBOAEMARQAiACwAIABpAGQAOgAgADMAIAB9ACwACgAJAHsAIABuAGEAbQBlADoAIAAiAEMASQBTACIALAAgAGkAZAA6ACAANAAgAH0ALAAKAAkAewAgAG4AYQBtAGUAOgAgACIARgBJAFIAUwBUAF8ATwBSAEQARQBSACIALAAgAGkAZAA6ACAANQAgAH0ALAAKAAkAewAgAG4AYQBtAGUAOgAgACIAUgBFAFMASQBTAFQAQQBOAEMARQAiACwAIABpAGQAOgAgADYAIAB9ACwACgBdACAAYQBzACAAYwBvAG4AcwB0ADsACgAKAGYAdQBuAGMAdABpAG8AbgAgAGYAaQBuAGQARgBhAGMAdABpAG8AbgAoAGkAZAA6ACAAbgB1AG0AYgBlAHIAKQAgAHsACgAJAHIAZQB0AHUAcgBuACAARgBBAEMAVABJAE8ATgBTAC4AZgBpAG4AZAAoACgAcAApACAAPQA%2BACAAcAAuAGkAZAAgAD0APQA9ACAAaQBkACkAPwAuAG4AYQBtAGUAIAA%2FAD8AIABuAHUAbABsADsACgB9AAoACgBlAHgAcABvAHIAdAAgAGYAdQBuAGMAdABpAG8AbgAgAGwAYQBiAGUAbAAoAGkAZAA6ACAAbgB1AG0AYgBlAHIAKQAgAHsACgAJAHMAdwBpAHQAYwBoACAAKABmAGkAbgBkAEYAYQBjAHQAaQBvAG4AKABpAGQAKQApACAAewAKAAkACQBjAGEAcwBlACAAIgBSAEUAUABVAEIATABJAEMAIgA6AAoACQAJAAkAcgBlAHQAdQByAG4AIAAiAEcAYQBsAGEAYwB0AGkAYwAgAFIAZQBwAHUAYgBsAGkAYwAiADsACgAJAAkAYwBhAHMAZQAgACIARQBNAFAASQBSAEUAIgA6AAoACQAJAAkAcgBlAHQAdQByAG4AIAAiAEcAYQBsAGEAYwB0AGkAYwAgAEUAbQBwAGkAcgBlACIAOwAKAAkACQBkAGUAZgBhAHUAbAB0ADoACgAJAAkACQByAGUAdAB1AHIAbgAgAG4AdQBsAGwAOwAKAAkAfQAKAH0A

### Expected result

No diagnostic.

### Code of Conduct

- [x] I agree to follow Biome's Code of Conduct

**Comment by ematipico** [comment]
You can work around the bug by adding explicit types 

```js
function findFaction(id: number): FACTIONS | null {
	return FACTIONS.find((p) => p.id === id)?.name ?? null;
}
```

**Comment by jakeleventhal** [comment]
I can still reproduce this on `@biomejs/biome@2.5.1` with a minimal project:

```jsonc
// biome.json
{
  "$schema": "https://biomejs.dev/schemas/2.5.1/schema.json",
  "linter": {
    "enabled": true,
    "rules": {
      "recommended": false,
      "suspicious": {
        "noUnnecessaryConditions": "error"
      }
    }
  }
}
```

```ts
// src/index.ts
type Movie = {
  characters?: Array<{
    isJedi: boolean;
    name: string;
  }>;
};

export function getJediNames(movie: Movie): Array<string> {
  const jedis = movie.characters?.filter((character) => character.isJedi) ?? [];

  return jedis.map((character) => character.name);
}
```

Steps:

```sh
bun install
bun run lint --colors=off
```

Observed:

```text
src/index.ts:9:16 lint/suspicious/noUnnecessaryConditions
  × Unnecessary nullish coalescing.
  i Drop ?? and the fallback expression.
  i The left-hand side is guaranteed to be non-nullish, so the fallback is unreachable.
```

Expected: no diagnostic. `movie.characters` is optional, so `movie.characters?.filter(...)` can evaluate to `undefined`; the `?? []` fallback is reachable and removing it changes the function's return type/behavior.

`bunx tsc --noEmit` passes under `strict: true`.

**Comment by johntalton** [comment]
Seeing the same issue here:
With the variation being "This comparison always has the same result." 

[Playground](https://next.biomejs.dev/playground/?lintRules=all&tab=formatter&pane=Diagnostics&code=LwAqACoAIABiAGkAbwBtAGUALQBpAGcAbgBvAHIAZQAtAGEAbABsACAAbABpAG4AdAAvAHMAdAB5AGwAZQAvAG4AbwBUAGUAcgBuAGEAcgB5ADoAIABiAGUAYwBhAHUAcwBlACAAKgAvAAoACgB0AHkAcABlACAAQgBsAGEAaAAgAD0AIAAnAGEAJwB8ACcAYwAnAAoAdAB5AHAAZQAgAEwAaQBzAHQATABpAHMAdABpAG4AZwAgAD0AIABSAGUAYwBvAHIAZAA8AHMAdAByAGkAbgBnAHwAQgBsAGEAaAAsACAAcwB0AHIAaQBuAGcAPgAKAAoAYwBvAG4AcwB0ACAATABJAFMAVAA6ACAATABpAHMAdABMAGkAcwB0AGkAbgBnACAAPQAgAHsAIABhADoAIAAnADEAJwAsACAAYgA6ACAAJwAyACcALAAgAGMAOgAgACcAMwAnACAAfQAKAAoAZgB1AG4AYwB0AGkAbwBuACAAaQBzAEsAZQB5ACgAawA6ACAAcwB0AHIAaQBuAGcAKQA6ACAAawAgAGkAcwAgAEIAbABhAGgAIAB7ACAAcgBlAHQAdQByAG4AIABrACAAIQA9AD0AIAAnAGIAJwAgAH0ACgAKAGYAdQBuAGMAdABpAG8AbgAgAHQAZQBzAHQAKABsAGkAcwB0AGkAbgBnADoAIABMAGkAcwB0AEwAaQBzAHQAaQBuAGcAKQA6ACAAQgBsAGEAaABbAF0ACgB7AAoAIAAgAHIAZQB0AHUAcgBuACAATwBiAGoAZQBjAHQALgBrAGUAeQBzACgAbABpAHMAdABpAG4AZwApAAoAIAAgACAAIAAuAG0AYQBwACgAawAgAD0APgAgAGkAcwBLAGUAeQAoAGsAKQAgAD8AIABrACAAOgAgAHUAbgBkAGUAZgBpAG4AZQBkACAAKQAKACAAIAAgACAALgBmAGkAbAB0AGUAcgAoAGsAIAA9AD4AIABrACAAIQA9AD0AIAB1AG4AZABlAGYAaQBuAGUAZAApACAALwAvACAAbgBvAFUAbgBuAGUAYwBlAHMAcwBhAHIAeQBDAG8AbgBkAGkAdABpAG8AbgBzAAoAfQAKAAoAdABlAHMAdAAoAEwASQBTAFQAKQAKAA%3D%3D&language=ts&ruleDomains.project=all)


Thanks
